infertrade/utilities/operations.py

```diff
@@ -374,45 +374,92 @@ def transform(self, X: pd.DataFrame, y=None) -> pd.DataFrame:
+) -> pd.DataFrame:
     """
-    This function is intended to be used as a decorator that can be applied to functions that calculate
-    allocation values.
-    This decorator takes two parameters: the first one is the lower limit for allocation values, and the second
-    is the upper limit for allocation values. Values that are below the lower limit or above the upper limit are
-    simply replaced by the lower and upper limits, respectively.
+    This function limits the ranges by limiting upper and lower limit of allocated values
 
     params:
+    allocated_dataframe
     allocation_lower_limit: the lower limit for allocation values.
     allocation_upper_limit: the upper limit for allocation values.
 
-    Example usage:
-    @limit_allocation(0, 2.5)
-    def myfunction(df: pandas.DataFrame) -> pandas.DataFrame:
-        return some_allocation_strategy(df)
+    returns:
+    allocation limited dataframe
     """
-
     if allocation_lower_limit > allocation_upper_limit:
         raise ValueError(
-                'The lower limit for allocation values should not be greater than the upper limit for'
-                ' allocation values.'
-                )
-
-    def wrapper(allocation_function: callable) -> callable:
-        def limited_function(*args, **kwargs) -> pd.DataFrame:
-            dataframe = allocation_function(*args, **kwargs)
-            dataframe.loc[
-                    dataframe[PandasEnum.ALLOCATION.value]
-                    > allocation_upper_limit, PandasEnum.ALLOCATION.value
-                    ] = allocation_upper_limit
-            dataframe.loc[
-                    dataframe[PandasEnum.ALLOCATION.value]
-                    <
-                    allocation_lower_limit, PandasEnum.ALLOCATION.value
-                    ] = allocation_lower_limit
-            return dataframe
-        return limited_function
-    return wrapper
+            "The lower limit for allocation values should not be greater than the upper limit for" " allocation values."
+        )
+    dataframe.loc[
+        dataframe[PandasEnum.ALLOCATION.value] > allocation_upper_limit, PandasEnum.ALLOCATION.value
+    ] = allocation_upper_limit
+    dataframe.loc[
+        dataframe[PandasEnum.ALLOCATION.value] < allocation_lower_limit, PandasEnum.ALLOCATION.value
+    ] = allocation_lower_limit
+    return dataframe
+
+
+def daily_stop_loss(dataframe: pd.DataFrame, loss_limit: float) -> pd.DataFrame:
+    """
+    This function calculates loss and limit the allocation accordingly.
+    It restricts allocation to 0 if loss>loss_limit
+
+    params:
+    allocated_dataframe
+    loss_limit
+
+    returns:
+    dataframe
+    """
+
+    prev_alloc = 0
+    prev_price = 0
+    for index, row in dataframe.iterrows():
+        price_change = row.price - prev_price
+        loss = -price_change * prev_alloc
+        if loss > loss_limit:
+            row.allocation = 0
+        prev_alloc = row.allocation
+        prev_price = row.price
```

Comment:
```suggestion
    prev_price = 0
    stop_loss_has_triggered = False
    for index, row in dataframe.iterrows():
        price_change = row.price - prev_price
        loss = -price_change * prev_alloc
        if stop_loss_has_triggered or loss > loss_limit:
            stop_loss_has_triggered = True
            row.allocation = 0
        prev_alloc = row.allocation
        prev_price = row.price
```

Reply:
There are different forms of stop loss, but a simple one could end the strategy going forward?

As discussed with @pawkw , you can also have a cumulative one, that checked if the cumulative loss exceeded an amount. You could also have a re-start point (e.g. 20 time steps later).